### PR TITLE
149 output utility

### PR DESCRIPTION
**Pull request by arichar6**
# Pull Request
## Description
Create new API and base class for OutputUtilities

This pull request addresses #149 

## Checklist
The following items have been checked for this PR:

- [ ] Conforms to PEP8 style standards (check with `pylint`, `flake8`, or similar)
- [x] Code is documented with docstrings, or docstrings have been updated as appropriate
- [ ] New unit test/integration test have been added to check new code
- [x] All existing tests still pass

### CODE REVIEW

File: turbopy/diagnostics.py [hunk 1]
```diff
@@ -109,6 +110,23 @@ def _write_buffer(self):
             np.savetxt(f, self._buffer, delimiter=",")
 
 
+
+class NetCDFOutputUtility(OutputUtility):
```
Comment (AOstenfeld) [inline turbopy/diagnostics.py:114]: Does this class need to override the `diagnose` and `finalize` methods from the `OutputUtility` base class?

**Review by AOstenfeld — CHANGES_REQUESTED** [review]

File: turbopy/diagnostics.py [hunk 2]
```diff
@@ -109,6 +110,23 @@ def _write_buffer(self):
...
+    def __init__(self, filename, diagnostic_size, **kwargs):
+        self._filename = filename
+        self._buffer = np.zeros(diagnostic_size)
+        self._buffer_index = 0
+
+
+utilities = {"stdout": PrintOutputUtility,
+             "csv": CSVOutputUtility,
+             "netcdf": NetCDFOutputUtility,
+            }
```
Comment (AOstenfeld) [inline turbopy/diagnostics.py:126]: Is this meant to be in the `NetCDFOutputUtility` class?

```suggestion
        utilities = {"stdout": PrintOutputUtility,
                     "csv": CSVOutputUtility,
                     "netcdf": NetCDFOutputUtility,
                    }
```

File: turbopy/diagnostics.py [hunk 3]
```diff
@@ -109,6 +110,23 @@ def _write_buffer(self):
...
+
+class NetCDFOutputUtility(OutputUtility):
+    """Diagnostic output utility that writes to NetCDF files
+    """
+    def __init__(self, filename, diagnostic_size, **kwargs):
+        self._filename = filename
+        self._buffer = np.zeros(diagnostic_size)
+        self._buffer_index = 0
+
+
```
Comment (AOstenfeld) [inline turbopy/diagnostics.py:122]: Does the `diagnose` method from `OutputUtility` need to be overridden?

```suggestion
    def diagnose(self, data):
        """Write data to NetCDF file"""
```

File: turbopy/diagnostics.py (hunk 2 above)
Comment (arichar6) [reply turbopy/diagnostics.py:126]: No, this is meant to be a "module level" variable, giving a mapping from the short name (the key) to the utility class. I could probably use the dynamic factory base class (that PhysicsModule etc use), but I thought that was unnecessarily complicated.

File: turbopy/diagnostics.py (hunk 3 above)
Comment (arichar6) [reply turbopy/diagnostics.py:122]: Yes. The class `NetCDFOutputUtility` is incomplete. I should probably just remove it.

**Comment by arichar6** [comment]
Thanks for the review. I've made a slight change, adding a new ABC method that should be useful for your PR #148. I've added `write_data` to the OutputUtility. You should be able to just call that method at some new interval. You don't have to worry about the details, as those will be taken care of in the concrete OutputUtility.

File: turbopy/diagnostics.py (hunk 3 above)
Comment (arichar6) [reply turbopy/diagnostics.py]: I've removed this class. We can add something like it later if desired.

**Comment by arichar6** [comment]
@AOstenfeld I'm going to go ahead and merge this, since I want to start using the HistoryDiagnostic for other things.

**Pull request merged**